src/mpid/ch4/netmod/ucx/ucx_progress.c

```diff
@@ -9,14 +9,10 @@
 
 #include "mpidimpl.h"
 #include "ucx_impl.h"
-//#include "events.h"
 
-static int am_handler(void *msg, size_t msg_sz);
-static void handle_am_recv(void *request, ucs_status_t status, ucp_tag_recv_info_t * info);
-
-static int am_handler(void *msg, size_t msg_sz)
+static void *am_buf;
+static void am_handler(void *request, ucs_status_t status, ucp_tag_recv_info_t * info)
```

Comment:
It is weird that there is no existing mechanism in ucp to pass down the buffer pointer to the call back function. What is in the `void *request`?

Reply:
That's the UCP request object allocated during `ucp_tag_msg_recv_nb`. We could store the buffer pointer in that object, but since we don't allocate it, there is race where the callback can execute before buffer location is set in the request.

A longer-term solution is to utilize the UCP active message APIs (currently experimental), which I believe could address this issue.

Reply:
I see. That makes sense. 

Could you add short comments explaining the lifecycle of `am_buf` and maybe a `TODO` on the future option of utilizing UCP active message APIs?